File: app/src/main/java/com/najmi/oreamnos/HashtagManagerActivity.java
package com.najmi.oreamnos;

import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;

import com.google.android.material.appbar.MaterialToolbar;
import com.google.android.material.chip.Chip;
import com.google.android.material.chip.ChipGroup;
import com.google.android.material.materialswitch.MaterialSwitch;
import com.google.android.material.textfield.TextInputEditText;

import com.najmi.oreamnos.utils.PreferencesManager;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Activity for managing hashtags that are appended to generated posts.
 */
public class HashtagManagerActivity extends AppCompatActivity {

    private static final String TAG = "HashtagManagerActivity";

    private PreferencesManager prefsManager;
    private MaterialSwitch autoAppendSwitch;
    private TextInputEditText hashtagInput;
    private TextView charCounter;
    private ChipGroup tagCloudChipGroup;
    private TextView clearHashtags;
    private TextView saveAction;

    private List<String> currentTags = new ArrayList<>();
    private boolean isLoading = true;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_hashtag_manager);

        prefsManager = new PreferencesManager(this);

        // Setup toolbar
        MaterialToolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        toolbar.setNavigationOnClickListener(v -> finish());

        // Initialize views
        autoAppendSwitch = findViewById(R.id.autoAppendSwitch);
        hashtagInput = findViewById(R.id.hashtagInput);
        charCounter = findViewById(R.id.charCounter);
        tagCloudChipGroup = findViewById(R.id.tagCloudChipGroup);
        clearHashtags = findViewById(R.id.clearHashtags);
        saveAction = findViewById(R.id.saveAction);

        // Load current settings
        loadSettings();

        // Setup listeners
        setupListeners();

        isLoading = false;
    }

    private void loadSettings() {
        // Load auto-append enabled state
        boolean hashtagsEnabled = prefsManager.areHashtagsEnabled();
        autoAppendSwitch.setChecked(hashtagsEnabled);

        // Load hashtags
        String hashtagsStr = prefsManager.getHashtags();
        if (hashtagsStr != null && !hashtagsStr.isEmpty()) {
            hashtagInput.setText(hashtagsStr);
            parseAndDisplayTags(hashtagsStr);
        }
    }

    private void setupListeners() {
        // Auto-append toggle
        autoAppendSwitch.setOnCheckedChangeListener((buttonView, isChecked) -> {
            if (!isLoading) {
                prefsManager.setHashtagsEnabled(isChecked);
                showSavedFeedback();
            }
        });

        // Hashtag input text changes
        hashtagInput.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
            }

            @Override
            public void afterTextChanged(Editable s) {
                String text = s.toString();
                updateCharCounter(text);
                if (!isLoading) {
                    parseAndDisplayTags(text);
                }
            }
        });

        // Clear hashtags button
        clearHashtags.setOnClickListener(v -> {
            hashtagInput.setText("");
            currentTags.clear();
            refreshTagCloud();
            if (!isLoading) {
                prefsManager.saveHashtags("");
                showSavedFeedback();
            }
        });

        // Save action
        saveAction.setOnClickListener(v -> {
            saveHashtags();
            Toast.makeText(this, "Hashtags saved", Toast.LENGTH_SHORT).show();
            finish();
        });

        // Add tag chip click
        Chip addTagChip = findViewById(R.id.addTagChip);
        if (addTagChip != null) {
            addTagChip.setOnClickListener(v -> {
                // Focus on input and show keyboard
                hashtagInput.requestFocus();
                hashtagInput.setSelection(hashtagInput.getText() != null ? hashtagInput.getText().length() : 0);
            });
        }
    }

    private void updateCharCounter(String text) {
        // Count hashtags, not characters
        String[] parts = text.split("[\\s,]+");
        int tagCount = 0;
        for (String part : parts) {
            if (!part.trim().isEmpty()) {
                tagCount++;
            }
        }
        charCounter.setText(tagCount + " / 30");
    }

    private void parseAndDisplayTags(String text) {
        currentTags.clear();

        if (text == null || text.trim().isEmpty()) {
            refreshTagCloud();
            return;
        }

        // Split by spaces or commas
        String[] parts = text.split("[\\s,]+");
        for (String part : parts) {
            String tag = part.trim();
            if (!tag.isEmpty()) {
                // Ensure tag starts with #
                if (!tag.startsWith("#")) {
                    tag = "#" + tag;
                }
                if (!currentTags.contains(tag)) {
                    currentTags.add(tag);
                }
            }
        }

        refreshTagCloud();
    }

    private void refreshTagCloud() {
        // Remove all chips except the Add Tag chip
        tagCloudChipGroup.removeAllViews();

        // Add chips for each tag
        for (String tag : currentTags) {
            Chip chip = new Chip(this);
            chip.setText(tag);
            chip.setTextColor(getColor(R.color.md_theme_light_onSurface));
            chip.setChipBackgroundColorResource(R.color.md_theme_light_surface);
            chip.setChipStrokeColorResource(R.color.md_theme_light_outline);
            chip.setChipStrokeWidth(getResources().getDimension(R.dimen.chip_stroke_width));
            chip.setCloseIconVisible(true);
            chip.setCloseIconTintResource(R.color.md_theme_light_onSurfaceVariant);

            // Handle chip close (remove tag)
            chip.setOnCloseIconClickListener(v -> {
                currentTags.remove(tag);
                refreshTagCloud();
                updateInputFromTags();
            });

            tagCloudChipGroup.addView(chip);
        }

        // Add the "Add Tag" chip at the end
        Chip addChip = new Chip(this);
        addChip.setId(R.id.addTagChip);
        addChip.setText("Add Tag");
        addChip.setTextColor(getColor(R.color.md_theme_light_onSurfaceVariant));
        addChip.setChipBackgroundColor(null);
        addChip.setChipStrokeColorResource(R.color.md_theme_light_outline);
        addChip.setChipStrokeWidth(getResources().getDimension(R.dimen.chip_stroke_width));
        addChip.setChipIconResource(R.drawable.ic_add);
        addChip.setChipIconTintResource(R.color.md_theme_light_onSurfaceVariant);
        addChip.setOnClickListener(v -> {
            hashtagInput.requestFocus();
            hashtagInput.setSelection(hashtagInput.getText() != null ? hashtagInput.getText().length() : 0);
        });
        tagCloudChipGroup.addView(addChip);
    }

    private void updateInputFromTags() {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < currentTags.size(); i++) {
            if (i > 0)
                sb.append(" ");
            sb.append(currentTags.get(i));
        }

        isLoading = true;
        hashtagInput.setText(sb.toString());
        isLoading = false;

        // Save after removing tag
        saveHashtags();
    }

    private void saveHashtags() {
        String hashtags = hashtagInput.getText() != null ? hashtagInput.getText().toString().trim() : "";
        prefsManager.saveHashtags(hashtags);
    }

    private void showSavedFeedback() {
        Toast.makeText(this, "Saved", Toast.LENGTH_SHORT).show();
    }

    @Override
    protected void onPause() {
        super.onPause();
        // Auto-save when leaving
        saveHashtags();
    }
}

File: app/src/main/java/com/najmi/oreamnos/HashtagManagerActivity.kt
package com.najmi.oreamnos

import android.os.Bundle
import android.widget.Toast
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.Close
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.focus.FocusRequester
import androidx.compose.ui.focus.focusRequester
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import com.najmi.oreamnos.ui.theme.SocurateTheme
import com.najmi.oreamnos.utils.PreferencesManager

/**
 * Activity for managing hashtags that are appended to generated posts.
 * Converted to Jetpack Compose.
 */
@OptIn(ExperimentalLayoutApi::class)
class HashtagManagerActivity : ComponentActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        
        val prefsManager = PreferencesManager(this)
        
        setContent {
            val currentTheme = prefsManager.getTheme()
            SocurateTheme(themeMode = currentTheme) {
                HashtagManagerScreen(
                    prefsManager = prefsManager,
                    onNavigateBack = { finish() }
                )
            }
        }
    }
}

@OptIn(ExperimentalMaterial3Api::class, ExperimentalLayoutApi::class)
@Composable
fun HashtagManagerScreen(
    prefsManager: PreferencesManager,
    onNavigateBack: () -> Unit
) {
    val context = LocalContext.current
    
    // State
    var hashtagsEnabled by remember { mutableStateOf(prefsManager.areHashtagsEnabled()) }
    var hashtagInput by remember { mutableStateOf(prefsManager.getHashtags() ?: "") }
    var currentTags by remember { mutableStateOf(parseTagsFromInput(hashtagInput)) }
    val focusRequester = remember { FocusRequester() }
    
    // Save on change
    fun saveHashtags() {
        prefsManager.saveHashtags(hashtagInput.trim())
    }
    
    Scaffold(
        topBar = {
            TopAppBar(
                title = { Text("Hashtag Manager") },
                navigationIcon = {
                    IconButton(onClick = { 
                        saveHashtags()
                        onNavigateBack() 
                    }) {
                        Icon(
                            imageVector = Icons.AutoMirrored.Filled.ArrowBack,
                            contentDescription = "Back"
                        )
                    }
                },
                actions = {
                    TextButton(onClick = {
                        saveHashtags()
                        Toast.makeText(context, "Hashtags saved", Toast.LENGTH_SHORT).show()
                        onNavigateBack()
                    }) {
                        Text("Save", color = MaterialTheme.colorScheme.primary)
                    }
                },
                colors = TopAppBarDefaults.topAppBarColors(
                    containerColor = MaterialTheme.colorScheme.background
                )
            )
        },
        containerColor = MaterialTheme.colorScheme.background
    ) { paddingValues ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(paddingValues)
                .verticalScroll(rememberScrollState())
                .padding(16.dp),
            verticalArrangement = Arrangement.spacedBy(24.dp)
        ) {
            // Auto-append toggle card
            Surface(
                modifier = Modifier.fillMaxWidth(),
                shape = RoundedCornerShape(16.dp),
                color = MaterialTheme.colorScheme.surface
            ) {
                Row(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(16.dp),
                    horizontalArrangement = Arrangement.SpaceBetween,
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    Column(modifier = Modifier.weight(1f)) {
                        Text(
                            text = "Auto-append hashtags",
                            style = MaterialTheme.typography.titleMedium,
                            color = MaterialTheme.colorScheme.onSurface
                        )
                        Spacer(modifier = Modifier.height(4.dp))
                        Text(
                            text = "Automatically add hashtags to generated posts",
                            style = MaterialTheme.typography.bodySmall,
                            color = MaterialTheme.colorScheme.onSurfaceVariant
                        )
                    }
                    Switch(
                        checked = hashtagsEnabled,
                        onCheckedChange = { enabled ->
                            hashtagsEnabled = enabled
                            prefsManager.setHashtagsEnabled(enabled)
                        },
                        colors = SwitchDefaults.colors(
                            checkedThumbColor = MaterialTheme.colorScheme.primary,
                            checkedTrackColor = MaterialTheme.colorScheme.primaryContainer
                        )
                    )
                }
            }
            
            // Hashtag input card
            Surface(
                modifier = Modifier.fillMaxWidth(),
                shape = RoundedCornerShape(16.dp),
                color = MaterialTheme.colorScheme.surface
            ) {
                Column(
                    modifier = Modifier.padding(16.dp),
                    verticalArrangement = Arrangement.spacedBy(16.dp)
                ) {
                    Row(
                        modifier = Modifier.fillMaxWidth(),
                        horizontalArrangement = Arrangement.SpaceBetween,
                        verticalAlignment = Alignment.CenterVertically
                    ) {
                        Text(
                            text = "Your Hashtags",
                            style = MaterialTheme.typography.titleMedium,
                            color = MaterialTheme.colorScheme.onSurface
                        )
                        Text(
                            text = "${currentTags.size} / 30",
                            style = MaterialTheme.typography.bodySmall,
                            color = MaterialTheme.colorScheme.onSurfaceVariant
                        )
                    }
                    
                    OutlinedTextField(
                        value = hashtagInput,
                        onValueChange = { newValue ->
                            hashtagInput = newValue
                            currentTags = parseTagsFromInput(newValue)
                        },
                        modifier = Modifier
                            .fillMaxWidth()
                            .focusRequester(focusRequester),
                        placeholder = {
                            Text(
                                text = "Enter hashtags separated by spaces",
                                color = MaterialTheme.colorScheme.onSurfaceVariant
                            )
                        },
                        minLines = 3,
                        maxLines = 5,
                        colors = OutlinedTextFieldDefaults.colors(
                            focusedBorderColor = MaterialTheme.colorScheme.primary,
                            unfocusedBorderColor = MaterialTheme.colorScheme.outline
                        )
                    )
                    
                    // Clear button
                    TextButton(
                        onClick = {
                            hashtagInput = ""
                            currentTags = emptyList()
                            prefsManager.saveHashtags("")
                        },
                        modifier = Modifier.align(Alignment.End)
                    ) {
                        Text(
                            text = "Clear All",
                            color = MaterialTheme.colorScheme.error
                        )
                    }
                }
            }
            
            // Tag cloud preview card
            Surface(
                modifier = Modifier.fillMaxWidth(),
                shape = RoundedCornerShape(16.dp),
                color = MaterialTheme.colorScheme.surface
            ) {
                Column(
                    modifier = Modifier.padding(16.dp),
                    verticalArrangement = Arrangement.spacedBy(12.dp)
                ) {
                    Text(
                        text = "Preview",
                        style = MaterialTheme.typography.titleMedium,
                        color = MaterialTheme.colorScheme.onSurface
                    )
                    
                    if (currentTags.isEmpty()) {
                        Text(
                            text = "No hashtags yet. Add some above!",
                            style = MaterialTheme.typography.bodyMedium,
                            color = MaterialTheme.colorScheme.onSurfaceVariant,
                            modifier = Modifier
                                .fillMaxWidth()
                                .padding(vertical = 16.dp),
                            textAlign = TextAlign.Center
                        )
                    } else {
                        FlowRow(
                            modifier = Modifier.fillMaxWidth(),
                            horizontalArrangement = Arrangement.spacedBy(8.dp),
                            verticalArrangement = Arrangement.spacedBy(8.dp)
                        ) {
                            currentTags.forEach { tag ->
                                HashtagChip(
                                    tag = tag,
                                    onRemove = {
                                        currentTags = currentTags.filter { it != tag }
                                        hashtagInput = currentTags.joinToString(" ")
                                        saveHashtags()
                                    }
                                )
                            }
                            
                            // Add tag chip
                            AddTagChip(onClick = { focusRequester.requestFocus() })
                        }
                    }
                }
            }
        }
    }
}

@Composable
fun HashtagChip(
    tag: String,
    onRemove: () -> Unit
) {
    Surface(
        shape = RoundedCornerShape(8.dp),
        color = MaterialTheme.colorScheme.surfaceVariant
    ) {
        Row(
            modifier = Modifier.padding(horizontal = 12.dp, vertical = 8.dp),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(4.dp)
        ) {
            Text(
                text = tag,
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurface
            )
            Icon(
                imageVector = Icons.Default.Close,
                contentDescription = "Remove",
                modifier = Modifier
                    .size(16.dp)
                    .clickable { onRemove() },
                tint = MaterialTheme.colorScheme.onSurfaceVariant
            )
        }
    }
}

@Composable
fun AddTagChip(onClick: () -> Unit) {
    Surface(
        modifier = Modifier.clickable { onClick() },
        shape = RoundedCornerShape(8.dp),
        color = MaterialTheme.colorScheme.background,
        border = ButtonDefaults.outlinedButtonBorder
    ) {
        Row(
            modifier = Modifier.padding(horizontal = 12.dp, vertical = 8.dp),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(4.dp)
        ) {
            Icon(
                imageVector = Icons.Default.Add,
                contentDescription = "Add",
                modifier = Modifier.size(16.dp),
                tint = MaterialTheme.colorScheme.onSurfaceVariant
            )
            Text(
                text = "Add Tag",
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
        }
    }
}

/**
 * Parse hashtags from input string
 */
private fun parseTagsFromInput(input: String): List<String> {
    if (input.isBlank()) return emptyList()
    
    return input.split(Regex("[\\s,]+"))
        .map { it.trim() }
        .filter { it.isNotEmpty() }
        .map { if (it.startsWith("#")) it else "#$it" }
        .distinct()
}


